hoprd/rest-api/src/session.rs

```diff
@@ -0,0 +1,131 @@
+use tokio_util::compat::FuturesAsyncReadCompatExt;
+use tracing::{error, info};
+
+use crate::{ApiErrorStatus, InternalState, BASE_PATH};
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
+#[schema(example = json!({
+        "destination": "12D3KooWR4uwjKCDCAY1xsEFB4esuWLF9Q5ijYvCjz5PNkTbnu33",
+        "path": {
+            "Hops": 1
+        },
+        "capabilities": ["Segmentation", "Retransmission"]
+    }))]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct SessionClientRequest {
+    #[serde_as(as = "DisplayFromStr")]
+    pub destination: PeerId,
+    pub path: PathOptions,
+    pub capabilities: Vec<SessionCapability>,
+    pub port: Option<u16>,
+}
+
+impl From<SessionClientRequest> for SessionClientConfig {
+    fn from(value: SessionClientRequest) -> Self {
+        Self {
+            peer: value.destination,
+            path_options: value.path,
+            capabilities: value.capabilities,
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
+#[schema(example = json!({
+        "port": 5542,
+    }))]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct SessionClientResponse {
+    pub port: u16,
+}
+
+/// Creates a new client session returing a dedicated session listening port.
+#[utoipa::path(
+        get,
+        path = const_format::formatcp!("{BASE_PATH}/session/client"),
+        request_body(
+            content = SessionClientRequest,
+            description = "Configuration of the client session to be created",
+            content_type = "application/json"),
+        responses(
+            (status = 200, description = "Successfully created a new client session", body = SessionClientResponse),
+            (status = 400, description = "Bad request.", body = ApiError),
+            (status = 401, description = "Invalid authorization token.", body = ApiError),
+            (status = 422, description = "Unknown failure", body = ApiError),
+            (status = 500, description = "Internal server error", body = ApiError)
+        ),
+        security(
+            ("api_token" = []),
+            ("bearer_token" = [])
+        ),
+        tag = "Messages"
+    )]
+pub(crate) async fn create_client(
+    State(state): State<Arc<InternalState>>,
+    Json(args): Json<SessionClientRequest>,
+) -> Result<impl IntoResponse, impl IntoResponse> {
+    let port = args.port.unwrap_or(0);
+    let data: SessionClientConfig = args.into();
+    let is_tcp_like = data.capabilities.contains(&SessionCapability::Retransmission)
+        || data.capabilities.contains(&SessionCapability::Segmentation);
+
+    if is_tcp_like {
+        let session = state.hopr.clone().connect_to(data).await.map_err(|e| {
+            (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                ApiErrorStatus::UnknownFailure(e.to_string()),
+            )
+        })?;
+
+        let tcp_listener = TcpListener::bind("127.0.0.1:{port}").await.map_err(|e| {
```

Comment:
Why are we using a port provided by the client instead of letting the OS choose a free port for us and send that back to the client?

Reply:
If you use a specific port, that one will be opened. If no value is supplied, 0 is used to request a random port and return that.